infra/presubmit.py

```diff
@@ -0,0 +1,308 @@
+
+import argparse
+import os
+import subprocess
+import sys
+import yaml
+
+_SRC_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+def _is_project_file(actual_path, expected_filename):
+  """Returns True if actual_path's name is |expected_filename| and is a file
+  that exists and is in in projects/."""
+  if os.path.basename(actual_path) != expected_filename:
+    return False
+
+  if os.path.basename(os.path.dirname(
+      os.path.dirname(actual_path))) != 'projects':
+    return False
+
+  return os.path.exists(actual_path)
+
+
+def _check_one_lib_fuzzing_engine(build_sh_file):
+  """Returns False if |build_sh_file| contains -lFuzzingEngine.
+  This is deprecated behavior. $LIB_FUZZING_ENGINE should be used instead
+  so that -fsanitize=fuzzer is used."""
+  if not _is_project_file(build_sh_file, 'build.sh'):
+    return True
+
+  with open(build_sh_file) as build_sh:
+    build_sh_lines = build_sh.readlines()
+  for line_num, line in enumerate(build_sh_lines):
+    uncommented_code = line.split('#')[0]
+    if '-lFuzzingEngine' in uncommented_code:
+      print('''Error: build.sh contains -lFuzzingEngine on line: {0}.
+Please use $LIB_FUZZING_ENGINE.'''.format(line_num))
+      return False
+  return True
+
+
+def check_lib_fuzzing_engine(paths):
+  """Call _check_one_lib_fuzzing_engine on each path in |paths|. Return True if
+  the result of every call is True."""
+  return all([_check_one_lib_fuzzing_engine(path) for path in paths])
+
+
+class ProjectYamlChecker:
+  """Checks for a project.yaml file."""
+
+  # Sections in a project.yaml and the constant values that they are allowed
+  # to have.
+  SECTIONS_AND_CONSTANTS = {
+      'sanitizers': {'address', 'none', 'memory', 'address'},
+      'architectures': {'i386', 'x86_64'},
+      'engines': {'afl', 'libfuzzer', 'honggfuzz'}
+  }
+
+  # Note: this list must be updated when we allow new sections.
+  VALID_SECTION_NAMES = [
+      'homepage', 'primary_contact', 'auto_ccs', 'sanitizers', 'architectures',
+      'disabled', 'view_restrictions', 'coverage_extra_args', 'vendor_ccs'
+  ]
+
+  # Note that some projects like boost only have auto-ccs. However, forgetting
+  # primary contact is probably a mistake.
+  REQUIRED_SECTIONS = ['primary_contact']
+
+  def __init__(self, filename):
+    self.filename = filename
+    with open(filename) as file_handle:
+      self.data = yaml.safe_load(file_handle)
+
+    self.success = True
+
+  def do_checks(self):
+    """Do all project.yaml checks. Return True if they pass."""
+    if self.is_disabled():
+      return True
+
+    checks = [
+        self.check_project_yaml_constants, self.check_required_sections,
+        self.check_valid_section_names, self.check_valid_emails
+    ]
+    for check_function in checks:
+      check_function()
+    return self.success
+
+  def is_disabled(self):
+    """Is this project disabled."""
+    return self.data.get('disabled', False)
+
+  def print_error_message(self, message, *args):
```

Comment:
nit: can shortify the name to just `error` or `log_error` or `print_error`, that would make the call sites below nicer

Reply:
Someone's been doing too much go programming :-)